packages/@ember/array/type-tests/index.ts

```diff
@@ -0,0 +1,214 @@
+import EmberArray, { A, NatifiedArray, isArray, makeArray } from '@ember/array';
+import MutableArray from '@ember/array/mutable';
+
+import { expectTypeOf } from 'expect-type';
+
+class Foo {
+  bar = 1;
+  hi(): string {
+    return 'Hi';
+  }
+  withArgs(foo: number, bar: string): string {
+    return `${foo}${bar}`;
+  }
+}
+
+let foo = new Foo();
+
+let arr = A([foo]);
+
+class Target {
+  foo = 2;
+}
+
+let target = new Target();
+
+expectTypeOf(arr).toMatchTypeOf<EmberArray<Foo>>();
+expectTypeOf(arr).toMatchTypeOf<MutableArray<Foo>>();
+expectTypeOf(arr).toEqualTypeOf<NatifiedArray<Foo>>();
+
+expectTypeOf(arr.length).toEqualTypeOf<number>();
+
+expectTypeOf(arr.objectAt(1)).toEqualTypeOf<Foo | undefined>();
+
+expectTypeOf(arr.firstObject).toEqualTypeOf<Foo | undefined>();
+
+expectTypeOf(arr.lastObject).toEqualTypeOf<Foo | undefined>();
+
+// The native version takes precedence
+expectTypeOf(arr.slice()).toEqualTypeOf<Foo[]>();
+expectTypeOf(arr.slice(1)).toEqualTypeOf<Foo[]>();
+expectTypeOf(arr.slice(1, 2)).toEqualTypeOf<Foo[]>();
+
+// The native version takes precedence
+expectTypeOf(arr.indexOf(new Foo())).toEqualTypeOf<number>();
+
+// The native version takes precedence
+expectTypeOf(arr.lastIndexOf(new Foo())).toEqualTypeOf<number>();
+
+expectTypeOf(arr.forEach((item: Foo) => String(item))).toEqualTypeOf<void>();
+
+arr.forEach((item, index, arr) => {
+  expectTypeOf(this).toEqualTypeOf<undefined>();
+  expectTypeOf(item).toEqualTypeOf<Foo>();
+  expectTypeOf(index).toEqualTypeOf<number>();
+  expectTypeOf(arr).toEqualTypeOf(arr);
+});
+arr.forEach(function (_item) {
+  // @ts-expect-error the EmberArray version isn't recognized
+  expectTypeOf(this).toEqualTypeOf(target);
+}, target);
+
+expectTypeOf(arr.getEach('bar')).toEqualTypeOf<NatifiedArray<number>>();
+expectTypeOf(arr.getEach('missing')).toEqualTypeOf<NatifiedArray<unknown>>();
+
+expectTypeOf(arr.setEach('bar', 2)).toEqualTypeOf(arr);
+// @ts-expect-error string is not assignable to bar
+arr.setEach('bar', 'string');
+
+// Can set unknown property
+expectTypeOf(arr.setEach('missing', 'anything')).toEqualTypeOf(arr);
+
+let mapped = arr.map((item, index, arr) => {
+  expectTypeOf(item).toEqualTypeOf<Foo>();
+  expectTypeOf(index).toEqualTypeOf<number>();
+  expectTypeOf(arr).toEqualTypeOf(arr);
+  return 1;
+});
+// NOTE: This should be a NatifiedArray, but the native version takes precedence
+expectTypeOf(mapped).toEqualTypeOf<number[]>();
```

Comment:
I think we might be able to do `toMatchTypeOf`?

```suggestion
expectTypeOf(mapped).toMatchTypeOf<NativeArray<number>>();
```

Reply:
Possibly. I'll review more.